File: Python3/187. Repeated DNA Sequences.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        patterns = collections.defaultdict(int)
        ans = []
        
        for i in range(9, len(s)):
            sub = s[i-9: i+1]
            if sub not in patterns:
                patterns[sub] = 1
            else:
                if patterns[sub] == 1:
                    ans.append(sub)
                    patterns[sub] += 1
                    
        
        return ans

File: Java/187. Repeated DNA Sequences.java
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        Map<String, Integer> patterns = new HashMap<>();
        List<String> ans = new ArrayList<>();
        
        for (int i=9; i<s.length(); i++) {
            String sub = s.substring(i-9, i+1);
            if (!patterns.containsKey(sub)) {
                patterns.put(sub, 1);
            } else {
                if (patterns.get(sub) == 1) {
                    ans.add(sub);
                    patterns.put(sub, 2);
                }
            }
        }
        
        return ans;
    }
}